source/retryable-writes/retryable-writes.rst

```diff
@@ -443,61 +443,65 @@ Consider the following pseudo-code:
+         * numbers" MUST be re-raised with an actionable error message.
+         */
+        if (!originalError.hasErrorLabel("RetryableWriteError")) {
+          if ( originalError.code == 20 && originalError.errmsg.startsWith("Transaction numbers") ) {
+            originalError.errmsg = "This MongoDB deployment does not support retryable...";
+          }
+          throw originalError
+        }
+
+        /*
+         * For exceptions that originate from the driver (e.g. no socket available
+         * from the connection pool), we should raise the previous error if there
+         * was one.
+         */
+        if (originalError is not DriverException && ! originalError.hasErrorLabel("NoWritesPerformed")) {
+          previousError = originalError;
         }
-        throw originalError
       }
-    }
 
-    /* If we cannot select a writable server, do not proceed with retrying and
-     * throw the original error. The caller can then infer that an attempt was
-     * made and failed. */
-    try {
-      server = selectServer("writable");
-    } catch (Exception ignoredError) {
-      throw originalError;
-    }
+      /* If we cannot select a writable server, do not proceed with retrying and
+       * throw the previous error. The caller can then infer that an attempt was
+       * made and failed. */
+      try {
+        server = selectServer("writable");
+      } catch (Exception ignoredError) {
+        throw previousError;
+      }
 
-    /* If the server selected for retrying is too old, throw the original error.
-     * The caller can then infer that an attempt was made and failed. This case
-     * is very rare, and likely means that the cluster is in the midst of a
-     * downgrade. */
-    if ( ! isRetryableWritesSupported()) {
-      throw originalError;
-    }
+      /* If the server selected for retrying is too old, throw the original error.
+       * The caller can then infer that an attempt was made and failed. This case
+       * is very rare, and likely means that the cluster is in the midst of a
+       * downgrade. */
+      if ( ! isRetryableWritesSupported(server)) {
+        throw originalError;
+      }
 
-    /* Allow any retryable error from the second attempt to propagate to our
-     * caller, as it will be just as relevant (if not more relevant) than the
-     * original error. For exceptions that originate from the driver (e.g. no
-     * socket available from the connection pool), we should raise the original
-     * error. Other exceptions originating from the server should be allowed to
-     * propagate. */
-    try {
-      return executeCommand(server, retryableCommand);
-    } catch (DriverException ignoredError) {
-      throw originalError;
-    } catch (Exception secondError) {
-      handleError(secondError);
-      throw secondError;
+      /* CSOT is enabled and the operation has timed out. */
+      if (timeoutMS != null && is_expired(timeoutMS) {
+        throw previousError;
```

Comment:
Do we have a spec test for this block?

Reply:
Yes we do, it's part of the CSOT. The "timeoutMS applies to whole operation, not individual attempts" tests: https://github.com/mongodb/specifications/blob/b3e06831a09ae72d94da79f428d8941ec7221f7f/source/client-side-operations-timeout/etc/templates/retryability-timeoutMS.yml.template#L42-L47